pkg/controller/compute/network.go

```diff
@@ -98,89 +94,115 @@ func (c *networkConnector) Connect(ctx context.Context, mg resource.Managed) (ma
-	cr, ok := mg.(*v1alpha3.Network)
+	cr, ok := mg.(*v1beta1.Network)
 	if !ok {
 		return managed.ExternalObservation{}, errors.New(errNotNetwork)
 	}
-	observed, err := c.Networks.Get(c.projectID, cr.Spec.Name).Context(ctx).Do()
-	if clients.IsErrorNotFound(err) {
-		return managed.ExternalObservation{
-			ResourceExists: false,
-		}, nil
-	}
+	observed, err := c.Networks.Get(c.projectID, meta.GetExternalName(cr)).Context(ctx).Do()
 	if err != nil {
-		return managed.ExternalObservation{}, err
+		return managed.ExternalObservation{}, errors.Wrap(resource.Ignore(gcp.IsErrorNotFound, err), errGetNetwork)
+	}
+
+	currentSpec := cr.Spec.ForProvider.DeepCopy()
+	network.LateInitializeSpec(&cr.Spec.ForProvider, *observed)
+	if !cmp.Equal(currentSpec, &cr.Spec.ForProvider) {
+		if err := c.kube.Update(ctx, cr); err != nil {
+			return managed.ExternalObservation{}, errors.Wrap(err, errManagedNetworkUpdate)
+		}
 	}
-	cr.Status.GCPNetworkStatus = network.GenerateGCPNetworkStatus(*observed)
-	// If the Network resource is retrieved, it is ready to be used
+
+	cr.Status.AtProvider = network.GenerateNetworkObservation(*observed)
+
 	cr.Status.SetConditions(runtimev1alpha1.Available())
+
+	u, _, err := network.IsUpToDate(meta.GetExternalName(cr), &cr.Spec.ForProvider, observed)
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(err, errCheckNetworkUpToDate)
+	}
+
 	return managed.ExternalObservation{
-		ResourceExists: true,
+		ResourceExists:   true,
+		ResourceUpToDate: u,
 	}, nil
 }
 
 func (c *networkExternal) Create(ctx context.Context, mg resource.Managed) (managed.ExternalCreation, error) {
-	cr, ok := mg.(*v1alpha3.Network)
+	cr, ok := mg.(*v1beta1.Network)
 	if !ok {
 		return managed.ExternalCreation{}, errors.New(errNotNetwork)
 	}
-	_, err := c.Networks.Insert(c.projectID, network.GenerateNetwork(cr.Spec.NetworkParameters)).
+
+	cr.Status.SetConditions(runtimev1alpha1.Creating())
+
+	net := &compute.Network{}
+	network.GenerateNetwork(meta.GetExternalName(cr), cr.Spec.ForProvider, net)
+	_, err := c.Networks.Insert(c.projectID, net).
 		Context(ctx).
 		Do()
-	if clients.IsErrorAlreadyExists(err) {
-		return managed.ExternalCreation{}, nil
-	}
-	cr.Status.SetConditions(runtimev1alpha1.Creating())
 	return managed.ExternalCreation{}, errors.Wrap(err, errNetworkCreateFailed)
 }
 
 func (c *networkExternal) Update(ctx context.Context, mg resource.Managed) (managed.ExternalUpdate, error) {
-	cr, ok := mg.(*v1alpha3.Network)
+	cr, ok := mg.(*v1beta1.Network)
 	if !ok {
 		return managed.ExternalUpdate{}, errors.New(errNotNetwork)
 	}
-	if cr.Spec.IsSameAs(cr.Status.GCPNetworkStatus) {
+
+	observed, err := c.Networks.Get(c.projectID, meta.GetExternalName(cr)).Context(ctx).Do()
+	if err != nil {
+		return managed.ExternalUpdate{}, errors.Wrap(resource.Ignore(gcp.IsErrorNotFound, err), errGetNetwork)
+	}
+
+	upToDate, switchToCustom, err := network.IsUpToDate(meta.GetExternalName(cr), &cr.Spec.ForProvider, observed)
```

Comment:
> You could possibly move the if statement that decides whether it should switch or not from IsUpToDate to here

I think this would violate the separation we want between the `client` and `controller` packages. If we did this, we would need to call `GenerateNetwork` in the controller then compare it to the observed, which is exactly what the `IsUpToDate` function is doing.

Reply:
Ah I see in your comment below that you want to compare to the actual `spec` value instead. Responded on that comment.